syntax = "proto3";

package stats;

message PostStatsRequest {
    string post_id = 1;
}

message PostStatsResponse {
    int64 views = 1;
    int64 likes = 2;
    int64 comments = 3;
}

message PostTimelineRequest {
    string post_id = 1;
    string start_date = 2;
    string end_date = 3;
}

message TimelineEntry {
    string date = 1;
    int64 count = 2;
}

message PostTimelineResponse {
    repeated TimelineEntry entries = 1;
}

message TopRequest {
    enum MetricType {
        VIEWS = 0;
        LIKES = 1;
        COMMENTS = 2;
    }

    MetricType metric_type = 1;
    int32 limit = 2;
}

message TopPostEntry {
    string post_id = 1;
    string title = 2;
    int64 count = 3;
}

message TopUserEntry {
    string user_id = 1;
    string username = 2;
    int64 count = 3;
}

message TopPostsResponse {
    repeated TopPostEntry posts = 1;
}

message TopUsersResponse {
    repeated TopUserEntry users = 1;
}

service StatsService {
    rpc GetPostStats(PostStatsRequest) returns (PostStatsResponse);
    rpc GetPostViewsTimeline(PostTimelineRequest) returns (PostTimelineResponse);
    rpc GetPostLikesTimeline(PostTimelineRequest) returns (PostTimelineResponse);
    rpc GetPostCommentsTimeline(PostTimelineRequest) returns (PostTimelineResponse);
    rpc GetTopPosts(TopRequest) returns (TopPostsResponse);
    rpc GetTopUsers(TopRequest) returns (TopUsersResponse);
}